### ISSUE TITLE
Update es6.d.ts for final draft of ES6 specification

### ISSUE THREAD

**Issue by rbuckton** [Microsoft/TypeScript#2881]
- [x] [6.1.5.1](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-well-known-symbols) - Add built in symbols to `SymbolConstructor`: `match`, `replace`, `search`, `species`, `split` (from #2524). 
- [x] [19.2.3](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-properties-of-the-function-prototype-object) - Remove `Function.toMethod`.
- [x] [19.2.3.6](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-function.prototype-@@hasinstance) - Add `[Symbol.hasInstance]` method to `Function` interface.
- [x] [20.3.4.45](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-date.prototype-@@toprimitive) - Add `[Symbol.toPrimitive]` method to `Date` interface.
- [x] [21.1.3.7](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-string.prototype.includes) - Add `includes` method to `String` interface, and remove `contains`
- [x] [21.1.3.11](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-string.prototype.match) - Add overload for `match` method on `String` interface that accepts an object with a compatible `[Symbol.match]` method.
- [x] [21.1.3.14](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-string.prototype.replace) - Add overload for `replace` method on `String` interface that accepts an object with a compatible `[Symbol.replace]` method.
- [x] [21.1.3.15](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-string.prototype.search) - Add overload for `search` method on `String` interface that accepts an object with a compatible `[Symbol.search]` method.
- [x] [21.1.3.17](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-string.prototype.split) - Add overload for `split` method on `String` interface that accepts an object with a compatible `[Symbol.split]` method.
- [x] [21.2.4.2](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-get-regexp-@@species) - Add `[Symbol.species]` property to `RegExpConstructor` interface.
- [x] [21.2.5.6](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-regexp.prototype-@@match) - Add `[Symbol.match]` method to `RegExp` interface.
- [x] [21.2.5.8](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-regexp.prototype-@@replace) - Add `[Symbol.replace]` method to `RegExp` interface.
- [x] [21.2.5.9](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-regexp.prototype-@@search) - Add `[Symbol.search]` method to `RegExp` interface.
- [x] [21.2.5.11](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-regexp.prototype-@@split) - Add `[Symbol.split]` method to `RegExp` interface.
- [x] [22.1.2.5](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-get-array-@@species) - Add `[Symbol.species]` property to `ArrayConstructor` interface.
- [x] [22.1.3.31](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-array.prototype-@@unscopables) - Add `[Symbol.unscopables]` property to `Array` interface.
- [x] [22.2.1.4](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-%typedarray%-object), [22.2.2.1.1](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-typedarrayfrom) - Add construct signature overload on Typed Array constructor interfaces to allow `Iterable<number>`.
- [x] [22.2.2.4](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-get-%typedarray%-@@species) - Add `[Symbol.species]` property to Typed Array constructor interfaces.
- [x] [22.2.3.31](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-get-%typedarray%.prototype-@@tostringtag) - Add `[Symbol.toStringTag]` property to Typed Array interfaces.
- [x] [23.1.2.2](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-get-map-@@species) - Add `[Symbol.species]` property to `MapConstructor` interface.
- [x] [23.2.2.2](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-get-set-@@species) - Add `[Symbol.species]` property to `SetConstructor` interface.
- [x] [24.1.3.3](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-get-arraybuffer-@@species) - Add `[Symbol.species]` property to `ArrayBufferConstructor` interface.
- [x] [25.2.3.3](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-generatorfunction.prototype-@@tostringtag) - Add `[Symbol.toStringTag]` property to `GeneratorFunction` interface.
- [x] [25.4.4.1](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-promise.all) - `Promise.all` can take an iterable.
- [x] [25.4.4.3](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-promise.race) - `Promise.race` can take an iterable.
- [x] [25.4.4.6](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-get-promise-@@species) - Add `[Symbol.species]` property to `PromiseConstructor` interface.
- [x] [25.4.5.4](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-promise.prototype-@@tostringtag) - Add `[Symbol.toStringTag]` property to `Promise` interface.
- [x] [26.1.2](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-reflect.construct) - Update `Reflect.construct` to add third `newTarget` argument.

**Comment by impinball** [comment]
And it's now the current standard.

**Comment by DanielRosenwasser** [comment]
@rbuckton it looks like `Promise.all` and `Promise.race` already take an iterable.

```TypeScript
interface PromiseConstructor {
    // ...

    /**
     * Creates a Promise that is resolved with an array of results when all of the provided Promises 
     * resolve, or rejected when any Promise is rejected.
     * @param values An array of Promises.
     * @returns A new Promise.
     */
    all<T>(values: Iterable<T | PromiseLike<T>>): Promise<T[]>;

    /**
     * Creates a Promise that is resolved or rejected when any of the provided Promises are resolved 
     * or rejected.
     * @param values An array of Promises.
     * @returns A new Promise.
     */
    race<T>(values: Iterable<T | PromiseLike<T>>): Promise<T>;

    // ...
}
```

**Comment by borekb** [comment]
Another one missing: `Array.prototype.includes` ([22.1.3.11](https://tc39.github.io/ecma262/2016/#sec-array.prototype.includes)).

**Comment by mhegazy** [comment]
> Another one missing: Array.prototype.includes (22.1.3.11).

This is not part of ES6, this is actually part of ES7/ES2016. the definition can be included using `--lib es2016`

**Comment by borekb** [comment]
Oh I see, thanks for the pointer!

**Comment by mhegazy** [comment]
Fixed by #18652
